data-serving/data-service/src/controllers/preprocessor.ts

```diff
@@ -0,0 +1,82 @@
+import { Case, CaseDocument } from '../model/case';
+import { NextFunction, Request, Response } from 'express';
+
+import { RevisionMetadata } from '../model/revision-metadata';
+
+const createNewMetadata = (curatorEmail: string, creationNotes: string) => {
+    return {
+        revisionNumber: 0,
+        creationMetadata: {
+            curator: curatorEmail,
+            date: Date.now(),
+            notes: creationNotes,
+        },
+    };
+};
+
+const createUpdateMetadata = (
+    c: CaseDocument,
+    curatorEmail: string,
+    updateNotes: string,
+) => {
+    return {
+        revisionNumber: ++c.revisionMetadata.revisionNumber,
+        creationMetadata: c.revisionMetadata.creationMetadata,
+        updateMetadata: {
+            curator: curatorEmail,
+            date: Date.now(),
+            notes: updateNotes,
+        },
+    };
+};
+
+export const setRevisionMetadata = async (
+    request: Request,
+    response: Response,
+    next: NextFunction,
+) => {
+    const curatorEmail = request.body.curator.email;
+    const caseReference = request.body?.caseReference;
+
+    // Find the case if it already exists so we can update its existing
+    // metadata.
+    let c;
+    if (request.method == 'PUT' && request.params?.id) {
+        // Update.
+        c = await Case.findById(request.params.id);
+    } else if (
+        request.method == 'PUT' &&
+        caseReference &&
+        caseReference.sourceId &&
+        caseReference.sourceEntryId
+    ) {
+        // Upsert.
+        // TODO: Upserts should only generate update metadata if there is a
+        // diff with what's already in the database.
+        c = await Case.findOne({
+            'caseReference.sourceId': caseReference.sourceId,
+            'caseReference.sourceEntryId': caseReference.sourceEntryId,
+        });
+    }
+
+    // Set the correct, server-generated revisionMetadata for subsequent
+    // processors to use.
+    const revisionMetadata = c
+        ? createUpdateMetadata(
+              c,
+              curatorEmail,
+              request.body?.revisionMetadata?.updateMetadata?.notes,
```

Comment:
Hm -- so we're now supplying a separate field for `curatorEmail`, but if there are curator notes, we're still expecting them to be populated in `revisionMetadata`? It's probably clearer _not_ to have `revisionMetadata` in the request type, here. Otherwise there are multiple places one might expect curator email to be drawn from.

Reply:
I knowww it's horrible.

We're not populating these notes so I can drop this piece of the PR until we figure out a better pattern, but see my comment above for why it is the way it is. LMK what you think.

Reply:
Dropping the notes is probably cool for now (don't think it has/will imminently come up in testing), but I think we ultimately want to have the field. Would it be workable just to put `notes` into `curator`, so that there's no expectation of `revisionMetadata` fields being necessary for a write? Sorry if I've mis-grok'd the comments above.

Reply:
I think that make sense. I've added a Trello card to circle back to these revision notes, but I think your suggestion is a good one. I think I need to understand where in the UX this will be surfaced (if anywhere).